dace/transformation/interstate/loop_to_map.py

```diff
@@ -202,7 +218,127 @@ def apply(self, sdfg: sd.SDFG):
+
+            # Create NestedSDFG and add all loop-body states and edges
+            # Also, find defined symbols in NestedSDFG
+            new_body = sdfg.add_state('single_state_body')
+            nsdfg = SDFG("loop_body", constants=sdfg.constants, parent=new_body)
+            nsdfg.add_node(body, is_start_state=True)
+            body.parent = nsdfg
+            exit_state = nsdfg.add_state('exit')
+            nsymbols = dict()
+            for state in states:
+                if state is body:
+                    continue
+                nsdfg.add_node(state)
+                state.parent = nsdfg
+            for state in states:
+                if state is body:
+                    continue
+                for src, dst, data in sdfg.in_edges(state):
+                    nsymbols.update({s: sdfg.symbols[s] for s in data.assignments.keys() if s in sdfg.symbols})
+                    nsdfg.add_edge(src, dst, data)
+            nsdfg.add_edge(body_end, exit_state, InterstateEdge())
+
+            # Move guard -> body edge to guard -> new_body
+            for src, dst, data, in sdfg.edges_between(guard, body):
+                sdfg.add_edge(src, new_body, data)
+            # Move body_end -> guard edge to new_body -> guard
+            for src, dst, data in sdfg.edges_between(body_end, guard):
+                sdfg.add_edge(new_body, dst, data)
+            
+            # Delete loop-body states and edges from parent SDFG
+            for state in states:
+                for e in sdfg.all_edges(state):
+                    sdfg.remove_edge(e)
+                sdfg.remove_node(state)
+            
+            # Add NestedSDFG arrays
+            for name in read_set | write_set:
+                nsdfg.arrays[name] = copy.deepcopy(sdfg.arrays[name])
+                nsdfg.arrays[name].transient = False
+            for name in unique_set:
+                nsdfg.arrays[name] = sdfg.arrays[name]
+                del sdfg.arrays[name]
+            
+            # Add NestedSDFG node
+            cnode = new_body.add_nested_sdfg(nsdfg, None, read_set, write_set)
+            if sdfg.parent:
+                for s, m in sdfg.parent_nsdfg_node.symbol_mapping.items():
+                    if s not in cnode.symbol_mapping:
+                        cnode.symbol_mapping[s] = m
+                        nsdfg.add_symbol(s, sdfg.symbols[s])
+            for name in read_set:
+                r = new_body.add_read(name)
+                new_body.add_edge(
+                    r, None, cnode, name,
+                    memlet.Memlet.from_array(name, sdfg.arrays[name]))
+            for name in write_set:
+                w = new_body.add_write(name)
+                new_body.add_edge(
+                    cnode, name, w, None,
+                    memlet.Memlet.from_array(name, sdfg.arrays[name]))
+
+            # Fix SDFG symbols
+            for sym, dtype in nsymbols.items():
+                if sym in sdfg.free_symbols:
```

Comment:
I would first collect the prior `sdfg.free_symbols` above this whole thing, and then check for the new ones (`sdfg.free_symbols - old_freesyms`). This is much safer w.r.t. actually free symbols.